File: Recursion/GFG/Medium/RatInMaze/RatInMaze.cpp
// https://www.geeksforgeeks.org/problems/rat-in-a-maze-problem/1
#include<bits/stdc++.h>
using namespace std;
void travel(vector<vector<int>> &mat, int i, int j, vector<string> &paths, string path)
{
    if (i < 0 || i >= mat.size() || j < 0 || j >= mat[0].size() || mat[i][j] == 0)
    {
        return;
    }
    if (i == mat.size() - 1 && j == mat[0].size() - 1)
    {
        paths.push_back(path);
        return;
    }
    mat[i][j] = 0; // mark visited
    // down
    travel(mat, i + 1, j, paths, path + "D");
    // up
    travel(mat, i - 1, j, paths, path + "U");
    // left
    travel(mat, i, j - 1, paths, path + "L");
    // right
    travel(mat, i, j + 1, paths, path + "R");
    mat[i][j] = 1; // unmark
}

vector<string> findPath(vector<vector<int>> &mat)
{
    // Your code goes here
    vector<string> paths;
    travel(mat, 0, 0, paths, "");
    return paths;
}
int main(void) {
    vector<vector<int>> mat = { {1, 0, 0, 0},
                                {1, 1, 0, 1},
                                {1, 1, 0, 0},
                                {0, 1, 1, 1} };
    vector<string> paths = findPath(mat); 
    for(string ele: paths) {
        cout << ele << " ";
    }
    return 0;
}

File: Recursion/GFG/Medium/RatInMaze/RatInMaze.java
// https://www.geeksforgeeks.org/problems/rat-in-a-maze-problem/1
import java.util.ArrayList;

public class RatInMaze {
    static void travel(int[][] mat, int i, int j, ArrayList<String> paths, String path) {
        if(i < 0 || i >= mat.length || j < 0 || j >= mat[0].length || mat[i][j] == 0) {
            System.out.println("Reached :" +i+","+j);
            return;
        }
        if(i == mat.length - 1 && j == mat[0].length - 1) {
            paths.add(path);
            return;
        }
        mat[i][j] = 0; // mark visited
        // down
        travel(mat, i+1, j, paths, path+"D");
        // right
        travel(mat, i, j+1, paths, path+"R");
        // left
        travel(mat, i, j-1, paths, path+"L");
        // up
        travel(mat, i-1, j, paths, path+"U");
        
        mat[i][j] = 1; // unmark
    }
    public static ArrayList<String> findPath(int[][] mat) {
        ArrayList<String> paths = new ArrayList<>();
        travel(mat, 0, 0, paths, "");
        return paths;
    }
    public static void main(String[] args) {
        int mat[][] = {{1, 0, 0, 0},
                {1, 1, 0, 1}, 
                {1, 1, 0, 0},
                {0, 1, 1, 1}};
        System.out.println(findPath(mat));
    }
}
